zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,355 @@
+            else:
+                msg = 'Resource type list is not a list!'
+                logging.error('{}'.format(msg))
+                raise
+            if len(types) > 0:
+                logging.info('Resource type list is populated '
+                             '({}, ok).'.format(len(types)))
+            else:
+                msg = 'Resource type list length is zero!'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Resource type list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_402_heat_stack_list(self):
+        """Check default heat stack list behavior, confirm functionality."""
+        logging.info('Checking default heat stack list...')
+        try:
+            stacks = list(self.heat_client.stacks.list())
+            if type(stacks) is list:
+                logging.info("Stack list check is ok.")
+            else:
+                msg = 'Stack list returned something other than a list.'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Heat stack list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_410_heat_stack_create_delete(self):
+        """Create stack, confirm nova compute resource, delete stack."""
+        logging.info('Creating, deleting heat stack (compute)...')
+        self._image_create()
+        self._keypair_create()
+        self._stack_create()
+        self._stack_resource_compute()
+        self._stack_delete()
+        self._image_delete()
+        self._keypair_delete()
+
+    def test_500_auth_encryption_key_same_on_units(self):
+        """Test the auth_encryption_key in heat.conf is same on all units."""
+        logging.info("Checking the 'auth_encryption_key' is the same on "
+                     "all units.")
+        output, ret = self._run_arbitrary(
+            "--application heat "
+            "--format json "
+            "grep auth_encryption_key /etc/heat/heat.conf")
+        if ret:
+            msg = "juju run returned error: ({}) -> {}".format(ret, output)
+            logging.error("Error: {}".format(msg))
+        output = json.loads(output)
+        keys = {}
+        for r in output:
+            k = r['Stdout'].split('=')[1].strip()
+            keys[r['UnitId']] = k
+        # see if keys are different.
+        ks = list(keys.values())
+        if any(((k != ks[0]) for k in ks[1:])):
+            msg = ("'auth_encryption_key' is not identical on every unit: {}"
+                   .format("{}={}".format(k, v) for k, v in keys.items()))
+            logging.error("Error: {}".format(msg))
+
+    @staticmethod
+    def _run_arbitrary(command, timeout=300):
+        """Run an arbitrary command (as root), but not necessarily on a unit.
+
+        (Otherwise the self.run(...) command could have been used for the unit
+
+        :param str command: The command to run.
```

Comment:
Please could we use the standard form for these.  i.e.

    :param command: The command to run
    :type command: str

Thanks.

Also, this function may raise subprocess.CalledProcessError and so also needs:

    :raises: subprocess.CalledProcessError.

Finally, the return type:

    :returns: A pair containing the output of the command and exit value
    :rtype: (str, int)

Reply:
Thanks @ajkavanagh. Admittedly, I copied & pasted this method from the Amulet implementation as-is.